src/check.py

```diff
@@ -0,0 +1,102 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Module for running checks."""
+
+import logging
+from collections.abc import Iterable, Iterator
+from typing import NamedTuple, TypeGuard
+
+from .content import conflicts as content_conflicts
+from .content import diff as content_diff
+from .types_ import AnyAction, UpdateAction
+
+
+class Problem(NamedTuple):
+    """Details about a failed check.
+
+    Attrs:
+        path: Unique identifier for the file and discourse topic with the problem
+        description: A summary of what the problem is and how to resolve it.
+
+    """
+
+    path: str
+    description: str
+
+
+def _is_update_action(action: AnyAction) -> TypeGuard[UpdateAction]:
+    """Check whether an action is an UpdateAction.
+
+    Args:
+        action: The action to check.
+
+    Returns:
+        Whether the action is an UpdateAction.
+    """
+    return isinstance(action, UpdateAction)
+
+
+def _update_action_problem(action: UpdateAction) -> Problem | None:
+    """Get any problem with an update action.
+
+    Args:
+        action: The action to check.
+
+    Returns:
+        None if there is no problem or the problem if there is an issue with the action.
+    """
+    if action.content_change is None:
```

Comment:
This validation logic is the same as the one in actions.py:180

Reply:
It is similar with a different purpose. The validation logic here needs to be run *before* any of the actions are executed. In actions.py, it is a final safeguard before updating the content on discourse

Reply:
Considering this validation is executed prior to the the action, does it still make sense to have this scenario in `action._update`?
```
elif (
            action.content_change is not None
            and action.content_change.base is None
            and action.content_change.local != action.content_change.server
        ):
            result = types_.ActionResult.FAIL
            reason = BASE_MISSING_REASON
```
Looks redundant to me

Reply:
Yes it is still needed since that function accepts any `UpdateAction` and will execute it. That function has no guarantee that it won't get passed an `UpdateAction` that shouldn't actually be updated. There is an edge case that will involve retrieving the content from discourse just before action execution and checking it again so I don't mind having the logic there

Reply:
Resolving for now, please let me know if you have further feedback on this